src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentNavigationService.cs

```diff
@@ -176,12 +176,10 @@ public bool TryNavigateToPosition(Workspace workspace, DocumentId documentId, in
             return NavigateTo(textBuffer, vsTextSpan);
         }
 
-        private static Document OpenDocument(Workspace workspace, DocumentId documentId, bool usePreviewTab)
+        private static Document OpenDocument(Workspace workspace, DocumentId documentId, bool reopenDocument, bool usePreviewTab)
         {
-            // Always open the document again, even if the document is already open in the 
-            // workspace. If a document is already open in a preview tab and it is opened again 
-            // in a permanent tab, this allows the document to transition to the new state.
-            if (workspace.CanOpenDocuments)
+            if (workspace.CanOpenDocuments &&
+                (reopenDocument || !workspace.IsDocumentOpen(documentId)))
```

Comment:
We already have an `INavigateToPreviewService` service which seems to be the "backdoor APIs NavigateTo needs to work in VS". Can we shove something there?

I'm proposing that OpenDocument should be rejected if it's already open. The workspace API there doesn't have any concept of tabs -- "reopen" is meaningless concept there.